posts/ci-test-with-github-actions-and-self-hosted-runners.md

```diff
@@ -0,0 +1,343 @@
+        ARMBIAN_BOARD=${board}
+        _info "ARMBIAN_BOARD=${board}"
+      fi
+    done
+  
+  done
+  _info "${ARMBIAN_BOARD} ${ARMBIAN_BRANCH} selected"
+}
+
+```
+
+Once the target is identified, the build is executed.
+
+```bash
+build_kernel() {
+  local build_config=${1}
+  ./compile.sh ${build_config} BOARD=${ARMBIAN_BOARD} BRANCH=${ARMBIAN_BRANCH}
+}
+```
+
+# Managing Secrets
+
+Our next phase of CI/CD Jobs involves GPG signing packages and uploading artifacts to our mirrors via rsync-over-ssh. 
+
+Github's access control around secrets isn't very robust. Any user with commit access to master branch can retrieve Github secrets.  This made storing our private keys directly in Github secrets unappealing.
+
+Since we consider our runners to be untrusted hosts, having the secrets stored on the filesystem was also unappealing.
+
+The compromise was to encrypt the secrets with ccrypt and store the encrypted filese on runners selected as designated hosts for uploading and signing. The decryption keyword is stored in Github secrets. A little bash scripting magic lets us only store the deycrypted values in memory, and never on disk--or on Github.
+
+`eval $(ssh-agent)` causes the script to launch ssh-agent and automatically export the environment variables ssh-agent outputs during runtime.  These values are needed for an ssh client to use the agent.
+
+`ssh-add <(ccat $PWD/${SSH_KEYFILE} -K ${{secrets.keyword}})` decrypts the ssh private key into memory using the github secret `keyword`.
+
+The `${{secret.variable}}` causes Github actions (not bash) to perform a substitution during run time.  The secret is automatically obfuscated by Github actions. 
+
+The `<(command)` redirects the output of a command into a file descriptor so a command can read it as a file stream, despite only being in memory.  Additionally ssh-agent only stores private keys in memory.
+
+Rsync is uploads the testfile via ssh because of the `user@host:` syntax. The ssh-private key is automatically used from ssh agent.
+
+```yaml
+jobs:
+  test_secrets_management:
+    name: do secrety things
+    runs-on: [self-hosted]
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
+          path: ci-test
+          clean: true
+      - name: Do some tests
+        shell: bash {0}
+        run: |
+         eval $(ssh-agent)
+         cd ci-test
+         TEST_HOST=testhost.example.com
+         TEST_USER=rsync_user
+         SSH_KEYFILE="secrets/test_ssh_key.cpt"
+         ssh-add <(ccat $PWD/${SSH_KEYFILE} -K ${{secrets.keyword}})
+         echo "Unique Data $(uuidgen)" > testfile
+         rsync -auv testfile ${TEST_USER}@${TEST_HOST}:
+```
+
+# Observations and Challenges
+
+
+## A workflow must exist before a worker can be added
+
+The Github Runner ansible role was unable to a runner to a repo that had no actions configured. First create a simple github actions workflow and commit to repo.  Once a basic workflow exist, a runner can be added.
```

Comment:
Missing a word I think

Reply:
fixed with word The Github Runner ansible role was unable register to a repo that had no actions configured.